segmentstore/server/src/test/java/io/pravega/segmentstore/server/writer/StorageWriterTests.java

```diff
@@ -232,6 +234,49 @@ public void testWithStorageTransientErrors() throws Exception {
      */
     @Test
     public void testWithStorageCorruptionErrors() throws Exception {
+        AtomicBoolean corruptionHappened = new AtomicBoolean();
+        Function<TestContext, ErrorInjector<Exception>> createErrorInjector = context -> {
+            byte[] corruptionData = "foo".getBytes();
+            SegmentHandle corruptedSegmentHandle = InMemoryStorage.newHandle(context.metadata.getStreamSegmentMetadata(0).getName(), false);
+            Supplier<Exception> exceptionSupplier = () -> {
+                // Corrupt data. We use an internal method (append) to atomically write data at the end of the segment.
+                // GetLength+Write would not work well because there may be concurrent writes that modify the data between
+                // requesting the length and attempting to write, thus causing the corruption to fail.
+                // NOTE: this is a synchronous call, but append() is also a sync method. If append() would become async,
+                // care must be taken not to block a thread while waiting for it.
+                context.storage.append(corruptedSegmentHandle, new ByteArrayInputStream(corruptionData), corruptionData.length);
+
+                // Return some other kind of exception.
+                return new TimeoutException("Intentional");
+            };
+            return new ErrorInjector<>(c -> !corruptionHappened.getAndSet(true), exceptionSupplier);
+        };
+
+        testWithStorageCriticalErrors(createErrorInjector, ex -> ex instanceof ReconciliationFailureException);
```

Comment:
I'm not sure how `TimeoutException` becomes `ReconciliationFailureException` for this scenario, could you elaborate?

Reply:
TimeoutException is thrown after we wrote garbage data to Tier 2.
The StorageWriter will retry the error.
Since it will pass the same offset, it will get `BadOffsetException`.
Since it gets `BadOffsetException`, it enters `ReconciliationMode.
Since we wrote garbage data which does not match what it should have been, the `StorageWriter` is failed with `ReconciliationFailureException`, which is a flavor of `DataCorruptionException`.